Source/ARTRealtimeChannel.m

```diff
@@ -276,21 +268,84 @@ - (void)publishProtocolMessage:(ARTProtocolMessage *)pm callback:(void (^)(ARTSt
-            } else {
+                [self sendMessage:pm callback:^(ARTStatus *status) {
+                    cb(status.errorInfo);
+                }];
+            }
+            else {
                 [self addToQueue:pm callback:queuedCallback];
 
-                [self.realtime.internalEventEmitter once:[ARTEvent newWithConnectionEvent:ARTRealtimeConnectionEventConnected] callback:^(ARTConnectionStateChange *__art_nullable change) {
+                [self.realtime.internalEventEmitter once:[ARTEvent newWithConnectionEvent:ARTRealtimeConnectionEventConnected] callback:^(ARTConnectionStateChange *_Nullable change) {
+                    [weakSelf sendQueuedMessages];
+                }];
+            }
+            break;
+        }
+    }
+} ART_TRY_OR_MOVE_TO_FAILED_END
+}
+
+- (void)publishProtocolMessage:(ARTProtocolMessage *)pm callback:(void (^)(ARTStatus *))cb {
+ART_TRY_OR_MOVE_TO_FAILED_START(_realtime) {
+    ARTStatus *statusInvalidConnectionState = [ARTStatus state:ARTStateError info:[ARTErrorInfo createWithCode:90001 message:@"channel operation failed (invalid connection state)"]];
+    ARTStatus *statusInvalidChannelState = [ARTStatus state:ARTStateError info:[ARTErrorInfo createWithCode:90001 message:@"channel operation failed (invalid channel state)"]];
+
+    __weak __typeof(self) weakSelf = self;
+    void (^queuedCallback)(ARTStatus *) = ^(ARTStatus *status) {
+        switch (weakSelf.state_nosync) {
+            case ARTRealtimeChannelSuspended:
+            case ARTRealtimeChannelFailed:
+                if (cb) {
+                    cb(status.state == ARTStateOk ? statusInvalidChannelState : status);
+                }
+                return;
+            default:
+                break;
+        }
+        if (cb) {
+            cb(status);
+        }
+    };
+
+    switch (_realtime.connection.state_nosync) {
+        case ARTRealtimeClosing:
+        case ARTRealtimeClosed: {
+            if (cb) {
+                cb(statusInvalidConnectionState);
+            }
+            return;
+        }
+        default:
+            break;
+    }
+
+    switch (self.state_nosync) {
+        case ARTRealtimeChannelSuspended:
+        case ARTRealtimeChannelFailed: {
+            if (cb) {
+                cb(statusInvalidChannelState);
+            }
+            break;
+        }
+        case ARTRealtimeChannelInitialized:
+        case ARTRealtimeChannelDetaching:
+        case ARTRealtimeChannelDetached:
+        case ARTRealtimeChannelAttaching:
+        case ARTRealtimeChannelAttached: {
+            if (_realtime.connection.state_nosync == ARTRealtimeConnected) {
+                [self sendMessage:pm callback:cb];
+            }
+            else {
+                [self addToQueue:pm callback:queuedCallback];
+                [self.realtime.internalEventEmitter once:[ARTEvent newWithConnectionEvent:ARTRealtimeConnectionEventConnected] callback:^(ARTConnectionStateChange *_Nullable change) {
```

Comment:
This seems unnecessary. Why have a channel-level queue for messages waiting for the connection to become connected (and manually listen for connection events) when the ARTRealtime class has its own queue that will do that automatically?

(Not to mention, seems like having a channel-level queue that mixes Messages waiting for connection CONNECTED with PresenceMessages waiting for channel ATTACHED is error-prone -- the event added here will drain the whole queue when the conneciton connects, including presence messages that actually need to wait for an attached, no?)

Reply:
Fixed a4466b7